File: 程序员面试金典/1.6字符串压缩/compressString.py
class Solution:
    def compressString(self, S: str) -> str:
        if not S:
            return ''
        compressed = []
        last = S[0]
        count = 1
        for i in range(1, len(S)):
            if S[i] == last:
                count += 1
            else:
                compressed.append(last + str(count))
                last = S[i]
                count = 1
        compressed.append(last + str(count))
        compressed_str = ''.join(compressed)
        return S if len(compressed_str) >= len(S) else compressed_str

File: 程序员面试金典/1.6字符串压缩/compressString.java
class Solution {
    public String compressString(String S) {
        if (S == null || S.length() <= 2)
            return S;
        StringBuilder compressed = new StringBuilder();
        char last = S.charAt(0);
        int count = 0;
        for (char c : S.toCharArray())
            if (c != last) {
                compressed.append(last).append(count);
                last = c;
                count = 1;
            }
            else
                ++count;
        compressed.append(last).append(count);
        return compressed.length() < S.length() ? compressed.toString() : S;
    }
}
